### PR TITLE
External package dependencies

### PR DESCRIPTION
**Pull request by paulcadman**
This PR adds external git dependency support to the Juvix package format.

## New dependency Git item

You can now add a `git` block to the dependencies list:

```yaml
name: HelloWorld
main: HelloWorld.juvix
dependencies:
  - .juvix-build/stdlib
  - git:
      url: https://my.git.repo
      name: myGitRepo
      ref: main
version: 0.1.0
```

Git block required fields:
* `url`: The URL of the git repository
* `ref`: The git reference that should be checked out
* `name`: The name for the dependency. This is used to name the directory of the clone, it is required. Perhaps we could come up with a way to automatically name the clone directory. Current ideas are to somehow encode the URL / ref combination or use a UUID. However there's some value in having the clone directory named in a friendly way.

NB:
* The values of the `name` fields must be unique among the git blocks in the dependencies list.

## Behaviour

When dependencies for a package are registered, at the beginning of the compiler pipeline, all remote dependencies are processed:

1. If it doesn't already exist, the remote dependency is cloned to `.juvix-build/deps/$name`
2. `git fetch` is run in the clone
3. `git checkout` at the specified `ref` is run in the clone

The clone is then processed by the PathResolver in the same way as path dependencies.

NB:
* Remote dependencies of transitive dependencies are also processed.
* The `git fetch` step is required for the case where the remote is updated. In this case we want the user to be able to update the `ref` field.

## Errors

1. Missing fields in the Git dependency block are YAML parse errors
2. Duplicate `name` values in the dependencies list is an error thrown when the package file is processed
3. The `ref` doesn't exist in the clone or the clone directory is otherwise corrupt. An error with a suggestion to `juvix clean` is given. The package file path is used as the location in the error message.
4. Other `git` command errors (command not found, etc.), a more verbose error is given with the arguments that were passed to the git command.

## Future work

1. Add an offline mode
2. Add a lock file mechanism that resolves branch/tag git refs to commit hashes

* closes https://github.com/anoma/juvix/issues/2083

### CODE REVIEW

File: tests/smoke/Commands/compile-dependencies.smoke.yaml [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+        git add -A
+        git commit -m "commit1"
+
+        dep1hash=$(git rev-parse HEAD)
+
+        # create project that uses dependency
+        mkdir $temp/base
+        cd $temp/base
+
+        cat <<-EOF > juvix.yaml
```
Comment (janmasrovira) [inline tests/smoke/Commands/compile-dependencies.smoke.yaml:33]: interesting, I didn't know about `cat <<-EOF`

**Review by janmasrovira — COMMENTED** [review]
Good job!

File: src/Juvix/Compiler/Concrete/Translation/FromParsed/Analysis/PathResolver.hs [hunk 2]
```diff
@@ -95,52 +106,77 @@ mkPackageInfo mpackageEntry _packageRoot = do
...
+getDependencyPath :: forall r. (Members '[Reader ResolverEnv, Files, Error DependencyError, GitClone] r) => PackageDependencyInfo -> Sem r (Path Abs Dir)
+getDependencyPath i = case i ^. packageDepdendencyInfoDependency of
+  DependencyPath p -> do
+    r <- asks (^. envRoot)
+    canonicalDir r (p ^. pathDependencyPath)
+  DependencyGit g -> do
+    r <- rootBuildDir <$> asks (^. envRoot)
+    let cloneDir = r <//> relDependenciesDir <//> relDir (T.unpack (g ^. gitDependencyName))
+    let cloneArgs = CloneArgs {_cloneArgsCloneDir = cloneDir, _cloneArgsRepoUrl = g ^. gitDependencyUrl}
+    let errorHandler' = errorHandler cloneDir
```
[the remark this answers is not in the document]
Comment (paulcadman) [reply src/Juvix/Compiler/Concrete/Translation/FromParsed/Analysis/PathResolver.hs]: Done

File: src/Juvix/Compiler/Pipeline/Package.hs [hunk 3]
```diff
@@ -148,25 +161,45 @@ processPackage buildDir pkg = do
         Right v -> return v
         Left err -> throw (pack (errorBundlePretty err))
 
+    checkNoDuplicateDepNames :: [Dependency] -> Sem r ()
+    checkNoDuplicateDepNames deps = go Set.empty (deps ^.. traversed . _GitDependency . gitDependencyName)
+      where
+        go :: Set Text -> [Text] -> Sem r ()
```
[the remark this answers is not in the document]
Comment (paulcadman) [reply src/Juvix/Compiler/Pipeline/Package.hs]: Now uses HashSet as suggested

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -151,6 +151,14 @@ jobs:
           extension-matching: disable
           cache: enable
 
+      # Smoke tests make git commits
+      - name: Setup git
+        shell: bash
+        run: |
+          git config --global user.email "tara-juvix@heliax.dev"
+          git config --global user.name "Tara"
```
[the remark this answers is not in the document]
Comment (paulcadman) [reply .github/workflows/ci.yml:159]: Yes :D. She recently replaced David Evnull <devull@heliax.dev> as director of communications.

**Review by janmasrovira — APPROVED** [review]

**Pull request merged**